virtcontainers/store/filesystem.go

```diff
@@ -0,0 +1,310 @@
+		"subsystem": "store",
+		"backend":   "filesystem",
+		"path":      f.path,
+	})
+}
+
+func (f *filesystem) trace(name string) (opentracing.Span, context.Context) {
+	if f.ctx == nil {
+		f.logger().WithField("type", "bug").Error("trace called before context set")
+		f.ctx = context.Background()
+	}
+
+	span, ctx := opentracing.StartSpanFromContext(f.ctx, name)
+
+	span.SetTag("subsystem", "store")
+	span.SetTag("type", "filesystem")
+	span.SetTag("path", f.path)
+
+	return span, ctx
+}
+
+func (f *filesystem) itemToPath(item Item) (string, error) {
+	fileName, err := itemToFile(item)
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(f.path, fileName), nil
+}
+
+func (f *filesystem) initialize() error {
+	_, err := os.Stat(f.path)
+	if err == nil {
+		return nil
+	}
+
+	// Our root path does not exist, we need to create the initial layout:
+	// The root directory, a lock file and a raw files directory.
+
+	// Root directory
+	if err := os.MkdirAll(f.path, DirMode); err != nil {
+		return err
+	}
+
+	// Raw directory
+	if err := os.MkdirAll(f.rawPath, DirMode); err != nil {
+		return err
+	}
+
+	// Lock
+	lockPath := filepath.Join(f.path, LockFile)
+
+	lockFile, err := os.Create(lockPath)
+	if err != nil {
+		f.delete()
+		return err
+	}
+	lockFile.Close()
+
+	return nil
+}
+
+func (f *filesystem) new(ctx context.Context, path string, host string) error {
+	f.ctx = ctx
+	f.path = path
+	f.rawPath = filepath.Join(f.path, "raw")
+	f.lockTokens = make(map[string]*os.File)
+
+	f.logger().Debugf("New filesystem store backend for %s", path)
+
+	return f.initialize()
+}
+
+func (f *filesystem) delete() {
+	os.RemoveAll(f.path)
+}
+
+func (f *filesystem) load(item Item, data interface{}) error {
+	span, _ := f.trace("load")
+	defer span.Finish()
+
+	span.SetTag("item", item)
+
+	filePath, err := f.itemToPath(item)
+	if err != nil {
+		return err
+	}
+
+	fileData, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(fileData, data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (f *filesystem) store(item Item, data interface{}) error {
+	span, _ := f.trace("store")
+	defer span.Finish()
+
+	span.SetTag("item", item)
+
+	filePath, err := f.itemToPath(item)
+	if err != nil {
+		return err
+	}
+
+	file, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	jsonOut, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("Could not marshall data: %s", err)
+	}
+	file.Write(jsonOut)
+
+	return nil
+}
+
+func (f *filesystem) raw(id string) (string, error) {
+	span, _ := f.trace("raw")
+	defer span.Finish()
+
+	span.SetTag("id", id)
+
+	// We must use the item ID.
+	if id != "" {
+		filePath := filepath.Join(f.rawPath, id)
+		file, err := os.Create(filePath)
+		if err != nil {
+			return "", err
+		}
+
+		return filesystemScheme + "://" + file.Name(), nil
+	}
+
+	// Generate a random temporary file.
+	file, err := ioutil.TempFile(f.rawPath, "raw-")
```

Comment:
`file` not closed.

Reply:
Renamed to `filesystem_backend.go`, thanks for the suggestion.